rfcs/2021-07-27-7117-vrl-string-interpolation.md

```diff
@@ -0,0 +1,188 @@
+# RFC 7117 - 2021-07-27 - VRL string interpolation
+
+VRL needs a better way to format strings. Currently the only way to do this is
+to concatenate strings, which can get unwieldly.
+
+## Scope
+
+This RFC discusses creating a new string type within VRL that can use template
+literals, also known as string interpolation, to format strings.
+
+## Pain
+
+Currently the way to create strings is through either string concatenation or
+the `join` function.
+
+Syntactically this is unwieldly. It requires extra key presses and the code
+created doesn't necessarily give an instant idea of what the resulting string
+will look like. Thus the true intent behind the code is obfuscated, which can
+result in bugs.
+
+## User Experience
+
+We will be loosely basing our format strings on Pythons [f-strings](https://peps.python.org/pep-0498/).
+
+To format a string there will be a string type denoted with the prefix `f'`.
+
+Within that string, it is possible to embed VRL expressions by surrounding them
+with `{..}`. VRL will evaluate the expressions and will call `as_string` on
+that expression to return the text representation.
+
+```coffee
+f'The message is { .message } created at { .timestamp }'
+```
+
+If you wish to actually insert a `{` into a string, a double '{{' will be needed.
+
+```coffee
+f'Here is a curly brace -> {{'
+```
+
+Since this is a new string type there are no backward compatibility issues.
+
+### Format strings
+
+The format can be specified by adding format strings after a `:` in the string.
+
+For example to format date the following would be valid:
+
+```coffee
+ts = t'2020-10-21T16:00:00Z'
+f'The time is {ts : %v %R}'
+```
+
+### Errors
+
+We do not want an f string to be fallible as this would cumbersome to the experience of using VRL.
+If an error occurs the error text is output. For example:
+
+```coffee
+f'This is some json { parse_json(.thing) }'
```

Comment:
Another alternative here would be to force users to handle errors explicitly with `??` or `parse_json!()` (which would force the program to abort). I slightly worry that users will neglect that the function is fallible and receive unexpected formatted output when it does fail.

Reply:
The way I have suggested" gives the user more choice. They can handle the error or just have the error output. If we forced them to handle the error they would only have the first option.

I'm not hugely opnionated on this and I can see that forcing them to handle errors explicitly would be more in keeping with the spirit of VRL - so happy to change this if you feel strongly?

Reply:
> They can handle the error or just have the error output.

That seems quite inconsistent with the error handling strategy we have right now. You'd be no longer able to scan your source for `abort` or `!` to see where a VRL script might error out.

Reply:
Another thought thrown into this: We could also make the format string fallible iff any of the expressions inside are fallible. So, in your case, handling it by:

```
f'This is some json { parse_json(.thing) }'!
```

Reply:
> We could also make the format string fallible iff any of the expressions inside are fallible.

I very much want to avoid this! I have updated it so each segment must be infallible.

Reply:
Could you elaborate what the benefits are of avoiding fallibility in this case? I can only see unexpected behavior at the cost of one saved character.